### DISCUSSION TITLE
looking for string->float function and how to get a word array from a string

### DISCUSSION THREAD

**Discussion by fnaj77** [janet-lang/janet#1226] [Q&A] [answered]
Hi guys, I can't find what I'm looking for in the docs so I'm here to ask for some help.

1) How can I convert a string to a number ("3.14" -> 3.14)?
2) Is there a simple way, ideally a single function, to extract the list of *non empty* words from a string? Something like what python would return for "a     b   c ".split() 

I'm quite confident that what I'm looking for is there in the docs, I can't just find it.

Additionally, does anyone know if any user has written a more organic manual than the official documentation and Ian's book "janet for mortals"?

Thank you for your help!

**Comment by tionis** [comment]
Regarding 1. you can use `(scan-number "3.14")`

**Comment by tionis** [comment]
For 2. perhaps [spork](https://github.com/janet-lang/spork)'s `sh/split` will do what you want?

**Comment by fnaj77** [comment]
Thank you very much @tionis, precious as always.

Regarding 1) I was searching with firefox the word "convert" in the full API list page. I should have thought about looking for parse/scan or similar words as well.

For 2), to be honest, I completely ignored spork's sh/* functions because I thought they were just shell related stuffs.

I'll wait a bit to mark this question as answered to see if anyone points me to a book or other material about janet.

Thank you again.

**Comment by CosmicToast** [comment, marked as answer]
1. `scan-number` works
2. use a peg, you don't even need to do post-processing.
```janet
# translation: skip whitespace (if any),
# then capture required non-whitespace characters,
# followed by any amount of whitespace characters,
# an unlimited amount of times
(def splitter '(* (any :s) (any (* '(some :S) (any :s)))))
(peg/match splitter "  ab   cd ") # => @["ab" "cd"]
```
3. Janet for mortals is about as good as you're going to get. I've been thinking of writing something too, but given my current situation that's not happening until at least next year.

**Reply by fnaj77** [reply]
Thank you @CosmicToast !

PEG is a super interesting feature of Janet and is something I've promised myself I'll look into as soon as I have more time. 
At the moment I was hoping to find a ready-made solution for something so common.
I'll probably end up blindly encapsulating your suggestion into a function to add to my own personal library, so I don't have to import spork just for sh/split.

As for the book, I figured there ware nothing else other than the main documentation and Ian's  and book. I gave it a try.
I'll wait and I'll certainly read your book as soon as it is ready.

Thank you again.

**Comment by pepe** [comment]
https://janet-lang.org/api/string.html#string/split for 2nd is possible too.

**Reply by fnaj77** [reply]
Hi @pepe.
I was actually looking for a function to extract the array of non empty words from a string.
With string/split, if there are more than one space between two contiguous words, you get a series of empty strings in between.

`repl:1:> (string/split " " "    a     b  c    ")`
`@["" "" "" "" "a" "" "" "" "" "b" "" "c" "" "" "" ""]`

With sh/split or with the peg based solution, you get an array with only non empty strings.

Thanks anyway for your reply.
